kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/VirtualCluster.java

```diff
@@ -5,20 +5,80 @@
+ * @param listeners virtual cluster listeners
+ * @param logNetwork if true, network will be logged
+ * @param logFrames if true, kafka rpcs will be logged
+ * @param filters filers.
+ */
+@SuppressWarnings("java:S1123") // suppressing the spurious warning about missing @deprecated in javadoc. It is the field that is deprecated, not the class.
 public record VirtualCluster(TargetCluster targetCluster,
-                             @JsonProperty(required = true) ClusterNetworkAddressConfigProviderDefinition clusterNetworkAddressConfigProvider,
-                             @JsonProperty() Optional<Tls> tls,
+                             @Deprecated(forRemoval = true, since = "0.11.0") ClusterNetworkAddressConfigProviderDefinition clusterNetworkAddressConfigProvider,
+                             @Deprecated(forRemoval = true, since = "0.11.0") @JsonProperty() Optional<Tls> tls,
+
+                             @JsonProperty(required = false) List<VirtualClusterListener> listeners,
                              boolean logNetwork,
                              boolean logFrames,
                              @Nullable List<String> filters) {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(VirtualCluster.class);
+
+    /**
+     * Name given to the listener defined using the deprecated fields.
+     */
+    @Deprecated(forRemoval = true, since = "0.11.0")
+    static final String DEFAULT_LISTENER_NAME = "default";
+
+    @SuppressWarnings({ "removal", "java:S2789" }) // S2789 - checking for null tls is the intent
+    public VirtualCluster {
+        if (clusterNetworkAddressConfigProvider != null || (tls != null && tls.isPresent())) {
+            if (clusterNetworkAddressConfigProvider == null) {
+                throw new IllegalConfigurationException("Deprecated virtualCluster property 'tls' supplied, but 'clusterNetworkAddressConfigProvider' is null");
+            }
+            if (listeners == null || listeners.isEmpty()) {
+                LOGGER.warn(
+                        "The virtualCluster properties 'clusterNetworkAddressConfigProvider' and 'tls' are deprecated, specify virtual cluster listeners using the listeners map.");
+                listeners = List.of(new VirtualClusterListener(DEFAULT_LISTENER_NAME, clusterNetworkAddressConfigProvider, tls));
+            }
+            else {
+                throw new IllegalConfigurationException(
+                        "When using listeners, the virtualCluster properties 'clusterNetworkAddressConfigProvider' and 'tls' must be omitted");
+            }
+        }
+        if (listeners == null || listeners.isEmpty()) {
+            throw new IllegalConfigurationException("no listeners configured for virtualCluster");
+        }
+        if (listeners.stream().anyMatch(Objects::isNull)) {
+            throw new IllegalConfigurationException("one or more listeners were null");
```

Comment:
My usual concern about blast radius, one miss configured listener ensures we can't start the proxy. Is that really what we want in this case? 

I know the more general fix to my concern is to make the config parsing/loading more robust and work at a per virtual cluster level but we don't live there yet. I can't make up my mind if this becomes more or less of an issue in a Kuernetes operator world. On the one hand new pods wont start if they are miss configured so the old pods will probably not leave service on the other hand the operator makes independent configuration of clusters a much more realistic thing and thus collisions more likely...

Reply:
Yeah, given operator validation I don't think there is anything for this or to do